Add GMMK Numpad (Default, Via)

RustedAperture opened the pull request:
<!--- Provide a general summary of your changes in the title above. -->

<!--- This template is entirely optional and can be removed, but is here to help both you and us. -->
<!--- Anything on lines wrapped in comments like these will not show up in the final text. -->

## Description

Adding the main framework needed to get QMK running along with Via and Vial. Most credit goes to @GloriousThrall.

## Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->
- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

tzarc commented:
I've gone ahead and unmarked the "I've read the [PR Checklist](https://docs.qmk.fm/#/pr_checklist)" as some things have been ignored.
Feel free to rectify the issues outlined in the document and let us know when you're done.

RustedAperture commented:
@tzarc I believe I've fixed the issues... I've removed Vial, added a readme, and moved the needed settings into info.json.
I've also removed duplicate and unneeded definitions in the config.h
If I'm missing anything please let me know

RustedAperture commented:
@adophoxia thank you for the guidance. I was basing my branch on the GMMK pro which still had the older header file containing the matrix. The above changes have been made.

adophoxia commented on keyboards/gmmk/numpad/info.json:
```diff
@@ -0,0 +1,54 @@
+    "layouts": {
+        "LAYOUT": {
+            "layout": [
+                {"label": "NUM", "x": 0, "y": 0, "matrix": [0, 0]},
+                {"label": "/", "x": 1, "y": 0, "matrix": [0, 1]},
+                {"label": "*", "x": 2, "y": 0, "matrix": [0, 2]},
+                {"label": "-", "x": 3, "y": 0, "matrix": [0, 3]},
+                
+                {"label": "7", "x": 0, "y": 1, "matrix": [1, 0]},
+                {"label": "8", "x": 1, "y": 1, "matrix": [1, 1]},
+                {"label": "9", "x": 2, "y": 1, "matrix": [1, 2]},
+                {"label": "+", "x": 3, "y": 1, "h": 2, "matrix": [1, 3]},
+
+                {"label": "4", "x": 0, "y": 2, "matrix": [2, 0]},
+                {"label": "5", "x": 1, "y": 2, "matrix": [2, 1]},
+                {"label": "6", "x": 2, "y": 2, "matrix": [2, 2]},
+                {"label": "CALC", "x": 3, "y": 2, "matrix": [2, 3]},
+
+                {"label": "1", "x": 0, "y": 3, "matrix": [3, 0]},
+                {"label": "2", "x": 1, "y": 3, "matrix": [3, 1]},
+                {"label": "3", "x": 2, "y": 3, "matrix": [3, 2]},
+                {"label": "RET", "x": 3, "y": 3, "h": 2, "matrix": [3, 3]},
+
+                {"label": "0", "x": 0, "y": 4, "w": 2, "matrix": [4, 0]},
+                {"label": ".", "x": 3, "y": 4, "matrix": [4, 3]}
```
```suggestion
                {"label": "NUM", "matrix": [0, 0], "x": 0, "y": 0},
                {"label": "/", "matrix": [0, 1], "x": 1, "y": 0},
                {"label": "*", "matrix": [0, 2], "x": 2, "y": 0},
                {"label": "-", "matrix": [0, 3], "x": 3, "y": 0},
                
                {"label": "7", "matrix": [1, 0], "x": 0, "y": 1},
                {"label": "8", "matrix": [1, 1], "x": 1, "y": 1},
                {"label": "9", "matrix": [1, 2], "x": 2, "y": 1},
                {"label": "+", "matrix": [1, 3], "x": 3, "y": 1, "h": 2},

                {"label": "4", "matrix": [2, 0], "x": 0, "y": 2},
                {"label": "5", "matrix": [2, 1], "x": 1, "y": 2},
                {"label": "6", "matrix": [2, 2], "x": 2, "y": 2},
                {"label": "CALC", "matrix": [2, 3], "x": 3, "y": 2},

                {"label": "1", "matrix": [3, 0], "x": 0, "y": 3},
                {"label": "2", "matrix": [3, 1], "x": 1, "y": 3},
                {"label": "3", "matrix": [3, 2], "x": 2, "y": 3},
                {"label": "RET", "matrix": [3, 3], "x": 3, "y": 3, "h": 2},

                {"label": "0", "matrix": [4, 0], "x": 0, "y": 4, "w": 2},
                {"label": ".", "matrix": [4, 3], "x": 3, "y": 4}
```

adophoxia commented on keyboards/gmmk/numpad/keymaps/default/keymap.c:
```diff
@@ -0,0 +1,56 @@
+};
+// clang-format on
+
+#ifdef ENCODER_ENABLE
+bool encoder_update_user(uint8_t index, bool clockwise) {
+    if (!clockwise) {
+      tap_code(KC_VOLU);
+    } else {
+      tap_code(KC_VOLD);
+    }
+    return true;
+}
+#endif
```
```suggestion

```
This isn't needed as this already exists in [`numpad.c`](https://github.com/qmk/qmk_firmware/pull/20067/files#diff-22bd728dacf18385a028533798f924c6eee6dbd7f17b2d3bd64b03711aa33af9R122-R134) at the keyboard level.

adophoxia commented on keyboards/gmmk/numpad/numpad.h:
```diff
@@ -0,0 +1,20 @@
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#pragma once
+
+#include "quantum.h"
+
+#define ___ KC_NO
+
+// clang-format on
```
```suggestion
#pragma once

#include "quantum.h"
```

adophoxia reviewed (approved):
LGTM! Better to remove the reference to `Vial` in the PR title. :)

RustedAperture commented:
> LGTM! Better to remove the reference to `Vial` in the PR title. :)

Removed, forgot I had it in there as I'm also working on vial at the same time. Thanks for the help!

waffle87 commented on keyboards/gmmk/numpad/config.h:
```diff
@@ -0,0 +1,86 @@
+
+#define DRIVER_COUNT 1
+
+#define DRIVER_1_LED_TOTAL 31
+#define RGB_MATRIX_LED_COUNT (DRIVER_1_LED_TOTAL)
+
```
```suggestion
#define RGB_MATRIX_LED_COUNT 31

```

waffle87 commented on keyboards/gmmk/numpad/keymaps/default/keymap.c:
```diff
@@ -0,0 +1,44 @@
+*/
+
+#include QMK_KEYBOARD_H
+#include "analog.h"
+
```
```suggestion
```

waffle87 commented on keyboards/gmmk/numpad/keymaps/via/keymap.c:
```diff
@@ -0,0 +1,63 @@
+*/
+
+#include QMK_KEYBOARD_H
+#include "analog.h"
```
```suggestion
```

waffle87 commented on keyboards/gmmk/numpad/numpad.c:
```diff
@@ -0,0 +1,134 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "numpad.h"
```
```suggestion
#include "quantum.h"
```

waffle87 commented on keyboards/gmmk/numpad/numpad.c:
```diff
@@ -0,0 +1,134 @@
+} };
+
+#    ifdef DRIVER_1_PW_EN
+void matrix_init_user(void) {
+    wait_ms(2000);
+    setPinOutput(DRIVER_1_PW_EN);
+    writePinHigh(DRIVER_1_PW_EN);
```
```suggestion
void matrix_init_kb(void) {
    wait_ms(2000);
    setPinOutput(DRIVER_1_PW_EN);
    writePinHigh(DRIVER_1_PW_EN);
    matrix_init_user();
```

waffle87 commented on keyboards/gmmk/numpad/numpad.c:
```diff
@@ -0,0 +1,134 @@
+    } else {
+        tap_code_delay(KC_VOLD, 10);
+    }
+    return false;
```
```suggestion
    return true;
```

waffle87 commented on keyboards/gmmk/numpad/numpad.c:
```diff
@@ -0,0 +1,134 @@
+    if (!encoder_update_user(index, clockwise)) {
+        return false;
+    }
+    if (!clockwise) {
```
This will produce odd behaviour, no?

RustedAperture replied:
> This will produce odd behaviour, no?

Not sure why @GloriousThrall had this, I removed it and it still works so gonna commit without the `encoder_update_user`

waffle87 replied:
No, that's not what I was talking about — the check for `encoder_update_user()` should be put back. I mean the `!clockwise` (not clockwise) -> volume up, clocwise -> volume down.

RustedAperture replied:
@waffle87 I put the `encoder_update_user()` back however I'm not sure what the issue is in this case... 
Clockwise = Volume up
Counter Clockwise = Volume Down
Thats what it is doing when I'm testing and what comes most naturally to me.

RustedAperture replied:
I believe the dial may be wired in reverse, I will do a bit more testing

RustedAperture replied:
@waffle87 above was the issue, the encoder pins are reversed, @GloriousThrall put the pins as A1, A2 meaning that we needed to have the encoder work in reverse.

waffle87 commented on keyboards/gmmk/numpad/numpad.h:
```diff
@@ -0,0 +1,16 @@
+/* Copyright 2021 Glorious, LLC <salman@pcgamingrace.com>
```
This file should be removed.

waffle87 commented on keyboards/gmmk/numpad/rules.mk:
```diff
@@ -0,0 +1,26 @@
+# MCU name
+MCU = WB32F3G71
+
+# Bootloader selection
+BOOTLOADER = wb32-dfu
+
```
```suggestion
```

RustedAperture commented:
Anything else I need to do or just waiting for merge now?

drashna commented on keyboards/gmmk/numpad/numpad.c:
```diff
@@ -0,0 +1,136 @@
+
+#    ifdef DRIVER_1_PW_EN
+
+void matrix_init_kb(void) {
+    wait_ms(2000);
+    setPinOutput(DRIVER_1_PW_EN);
+    writePinHigh(DRIVER_1_PW_EN);
+    matrix_init_user();
+}
```
```suggestion
void keyboard_post_init_kb(void) {
    wait_ms(2000);
    setPinOutput(DRIVER_1_PW_EN);
    writePinHigh(DRIVER_1_PW_EN);
    keyboard_post_init_user();
}
```

Also, is the wait needed, especially with the post init?

RustedAperture replied:
Not really sure, I can test and investigate in the morning to see if it is needed.

RustedAperture replied:
shouldn't this code be in `keyboard_pre_init_user` as per the documentation? The doc says led initialization should be there.
 https://github.com/qmk/qmk_firmware/blob/master/docs/custom_quantum_functions.md#example-keyboard_pre_init_user-implementation

drashna replied:
yeah, because wait_ms is hard stop (eg, nothing else runs).  So if it can be removed, that would be idea.

RustedAperture replied:
@drashna removing the `wait_ms(200);` breaks RGB, moving it to `keyboard_pre_init_user()` breaks RGB, moving it to `keyboard_post_init_kb()` also breaks RGB

Seems like the only way RGB still works is by keeping it in the matrix init code... executing it earlier than matrix breaks the RGB

RustedAperture replied:
@drashna I can move it to `keyboard_pre_init_user()` but it still requires the `wait_ms(200);`

sigprof replied:
A 2 second delay at startup is really unfortunate. At least it's a numpad, and not a regular keyboard.

Apparently `matrix_init()` is about the last replaceable hook that is invoked before `rgb_matrix_init()` (in theory there is also `led_init_ports()`, but that's for a completely different feature, so its usage may be considered as a dirty hack; and the only user-visible change would be that the 2 second delay would happen after the bootmagic check).

I wonder what is the underlying problem here — no sane chip would require such delays (the AW20216S datasheet mentions that a 2 ms delay is required after a power-on reset, but that's 1000 times lower than what is done here).

RustedAperture replied:
@sigprof I never tried to use a faster wait time, it could be that @GloriousThrall read 2 seconds which would be 2000 ms instead of 2 ms

sigprof replied:
If it was the case, it would have worked without the delay, because the USB init code has a 50 ms delay in it — but as you observe that RGB does not work without that delay, there is something else there that requires a long delay at startup.

RustedAperture replied:
I got this lowered to 1000ms, when I try to go lower than this it breaks RGB

sigprof replied:
Without any reliable information about why that delay is needed you may need to keep it at 2000 ms (maybe 1000 ms is enough for your particular board, but not in general).

Maybe the problem is caused by the wireless part (which is not touched by this code) — e.g., there might be some circuit which switches off the power to the LED controller.

BTW, one thing you can do here is to put the code inside `#ifdef RGB_MATRIX_ENABLE`, so that the delay won't be applied if someone decides to turn off RGB support completely.

RustedAperture replied:
@sigprof so I should change `#ifdef DRIVER_1_PW_EN` to `#ifdef RGB_MATRIX_ENABLE`

RustedAperture replied:
It looks like the entire RGB matrix along with this trouble code is already wrapped in `RGB_MATRIX_ENABLE`

sigprof replied:
Ah, then that's OK as is — did not look at the full file.

sigprof commented on keyboards/gmmk/numpad/readme.md:
```diff
@@ -0,0 +1,29 @@
+# GMMK NUMPAD
+
+A 17 Key macropad made ans old by Glorious LLC, equiped with a WB32 microcontroller, with support for a rotary encoder.
```
Looks like there are some typos here:
```suggestion
A 17 Key macropad made and sold by Glorious LLC, equipped with a WB32 microcontroller, with support for a rotary encoder.
```

sigprof commented on keyboards/gmmk/numpad/readme.md:
```diff
@@ -0,0 +1,29 @@
+
+* Keyboard Maintainer: [GloriousThrall](https://github.com/GloriousThrall)
+* Hardware Supported: GMMK Numpad
+* Hardware Availability: [GloriousPCGaming.com](https://www.gloriousgaming.com/products/gmmk-numpad)
```
That page says that GMMK Numpad supports both wired and wireless (Bluetooth 5.0 LE) connectivity; however, the code in this PR does not seem to have any kind of wireless support.  This loss of functionality should at least be noted in the readme.

sigprof commented on keyboards/gmmk/numpad/readme.md:
```diff
@@ -0,0 +1,29 @@
+
+## Bootloader
+
+Enter the bootloader in 3 ways:
+
+* **Bootmagic reset**: Hold down the Num key and plug in the keyboard
+* **Keycode in layout**: Press the Num+Del key which is mapped to `QK_BOOT`
```
Only 2 ways are mentioned here.  Does the PCB have any kind of a hardware reset button (for emergency usage, if some broken firmware version had been flashed)?

Also do the mentioned ways work when the board still has the Glorious Core firmware, or some other method needs to be used to flash the QMK firmware for the first time?

adophoxia replied:
There is a hardware reset button that's between the 2U `0` key and the `. Del` key on top of the PCB, which is accessible via a hole on the plate with a thin pin tool like a SIM card tool or paperclip.

RustedAperture commented:
The above commit moves the RGB to the `keyboard_pre_init_user()`. I tried to remove the `wait_ms(200);` and initialize the RGB in various different inits but it didn't work. keeping the `wait_ms(200);` is the only way the RGB doesn't break.

sigprof commented on keyboards/gmmk/numpad/readme.md:
```diff
@@ -0,0 +1,29 @@
+
+## Broken
+
+* Slider does not function
```
There are a couple of keyboards with sliders in QMK (`keebwerk/nano_slider`, `ungodly/launch_pad`); they map the slider to a MIDI controller input, which then can be used by some software on the host side.  However, I can understand that you may not want to enable MIDI support in the default firmware due to compatibility concerns.

One option is adding a separate keymap which supports the slider by mapping it to MIDI.

RustedAperture replied:
I've attempted this however I get an error regarding not enough endpoints.

sigprof replied:
Ugh, apparently WB32F3G71 has only 3 bidirectional endpoints, like STM32F4x1.  So you may need the same workaround if you want to fit VIA + MIDI there:
```make
KEYBOARD_SHARED_EP = yes
```
This may bring some compatibility issues though, because the 6KRO keyboard protocol will now go over the same endpoint as NKRO, media keys and mouse emulation.  However, many STM32F4x1-based boards use that option.

(No idea whether WB32F3G71 has other specific compatibility issues though — I don't have any devices with that chip.)

RustedAperture replied:
I will add it as a separate keymap for now, for some reason adding it to the default map will cause it to not build anymore.

RustedAperture replied:
I'm getting, even when I add analog.c and analog.h
```
| 
 | /usr/lib/gcc/arm-none-eabi/12.2.0/../../../../arm-none-eabi/bin/ld: /tmp/ccgN9wGW.ltrans0.ltrans.o: in function `matrix_scan.isra.0':
 | /home/cameronv/projects/qmk_rustedaperture/./keyboards/gmmk/numpad/keymaps/default/keymap.c:52: undefined reference to `analogReadPin'
 | collect2: error: ld returned 1 exit status
 | 
make[1]: *** [builddefs/common_rules.mk:267: .build/gmmk_numpad_default.elf] Error 1
make: *** [Makefile:392: gmmk/numpad:default] Error 1
Make finished with errors
```

sigprof replied:
For the analog support to work, you need to add `SRC += analog.c`, and also add `halconf.h` with `#define HAL_USE_ADC TRUE` and `mcuconf.h` with `#define WB32_ADC_USE_ADC1 TRUE` (you can find a sample in `keyboards/handwired/onekey/evb_wb32f3g71/`, although those files enable more peripherals).  Also you may need to run `qmk clean` after adding those files (the addition of extra files might not be properly picked up by the build system).

RustedAperture replied:
Thank you, got the slider working now.

sigprof commented on keyboards/gmmk/numpad/readme.md:
```diff
@@ -0,0 +1,29 @@
+## Broken
+
+* Slider does not function
+* Rotary input push button (mapped to gpio)
```
If the push switch connects that GPIO to GND, you can add support for it by declaring that GPIO pin as an extra row pin (because your matrix is `ROW2COL`, and therefore row pins are inputs), then using any matrix location in that row for the encoder key (connecting a matrix input to GND will result in all matrix locations in that row appearing as pressed).  You can also use the `MATRIX_MASKED` feature to mask all other locations in that row; this will avoid some useless keymap lookups when the encoder is pressed.

If the push switch is connected in some other way, at the moment the only solution to support it properly is implementing a custom matrix (which would unfortunately duplicate a lot of the standard matrix code).

RustedAperture replied:
I cannot find the pinout for the wb32f3g7. from using a continuity checker the rotary encoder sw and gnd are connected to 2 pins on the microprocessor

RustedAperture replied:
they appear to be commented to the pins directly beside the placement marker (dot on the Microcontroller)

adophoxia replied:
Refer to the [docs](https://github.com/qmk/qmk_firmware/pull/16871#issuecomment-1436188572) here.
* https://github.com/qmk/qmk_firmware/pull/16871#issuecomment-1436188572

RustedAperture replied:
@adophoxia is the pinout the same for the WB32F3G7 vs WB32FQ95?

RustedAperture replied:
@sigprof I've discovered that it is connected to VDD and PB8, when I add PB8 to the matrix row I get the following warning 
```
⚠ gmmk/numpad: led_config: numpad.c: Unable to parse g_led_config matrix data
```

RustedAperture replied:
Hub16 uses a custom matrix to get the encoder presses but this seems a little out of my scope of knowledge at the moment.

RustedAperture replied:
I got the rotary button working however it wont be customizable with via atm

adophoxia replied:
You can configure the VIA Definition JSON that it'll be used for the encoder with this guide [`here`](https://www.caniusevia.com/docs/layouts#optional-rotary-encoder-combined-switchrotary-encoder-footprint).

RustedAperture replied:
@adophoxia the issue is that the way I got it working was not adding it to the matrix. if you check the most recent commit you will understand.

RustedAperture replied:
rotary is added just not the push button, sorry should have specified. the push button is fixed as KC_MUTE since its not part of the matrix

RustedAperture commented:
Push button on encoder is now working, the idea to get it working came from [lynepad](https://github.com/qmk/qmk_firmware/tree/042e42d1b7de603bc0a164b7d3b2ec761c6e140f/keyboards/qvex/lynepad) however this comes with the trade-off of it not being customizable in via atm.

drashna commented on keyboards/gmmk/numpad/keymaps/default/keymap.c:
```diff
@@ -0,0 +1,70 @@
+extern int16_t enc;
+extern int16_t encPrev;
+
+void matrix_scan_user(void) {
```
```suggestion
void housekeeping_task_user(void) {
```

drashna commented on keyboards/gmmk/numpad/keymaps/via/keymap.c:
```diff
@@ -0,0 +1,90 @@
+extern int16_t enc;
+extern int16_t encPrev;
+
+void matrix_scan_user(void) {
```
```suggestion
void housekeeping_task_user(void) {
```

drashna commented on keyboards/gmmk/numpad/numpad.c:
```diff
@@ -0,0 +1,151 @@
+int16_t enc = 1;
+int16_t encPrev = 1;
+
+void matrix_scan_kb(void) {
```
```suggestion
void housekeeping_task_kb(void) {
```

drashna commented on keyboards/gmmk/numpad/numpad.c:
```diff
@@ -0,0 +1,151 @@
+void matrix_scan_kb(void) {
+    encPrev = enc;
+    enc = readPin(ENCODER_PUSHBUTTON_PIN);
+
+    matrix_scan_user();
```
```suggestion
```

RustedAperture commented:
Let me know if anything needs to be changed. Currently using this build daily without issues now.

drashna commented on keyboards/gmmk/numpad/info.json:
```diff
@@ -0,0 +1,54 @@
+            {"pin_a": "A2", "pin_b": "A1"}
+        ]
+    },
+    "bootmagic": {
+        "matrix": [0, 0]
+    },
```
This is the default location, so this shouldn't need to be specified, at all.
```suggestion
```

drashna commented on keyboards/gmmk/numpad/config.h:
```diff
@@ -0,0 +1,87 @@
+
+#pragma once
+
+#define TAP_CODE_DELAY 10
```
This could be moved to `post_config.h`, and added as:
```c
#ifndef TAP_CODE_DELAY
#    define TAP_CODE_DELAY 10
#endif
```

drashna commented on keyboards/gmmk/numpad/keymaps/default/keymap.c:
```diff
@@ -0,0 +1,71 @@
+    if (enc != encPrev) {
+        if (enc < 1) {
+          tap_code_delay(KC_MUTE, 10);
+        }
+        else {
```
```suggestion
```

drashna commented on keyboards/gmmk/numpad/keymaps/via/keymap.c:
```diff
@@ -0,0 +1,90 @@
+    if (enc != encPrev) {
+        if (enc < 1) {
+          tap_code_delay(KC_MUTE, 10);
+        }
+        else {
```
```suggestion
```

drashna commented on keyboards/gmmk/numpad/keymaps/via/keymap.c:
```diff
@@ -0,0 +1,90 @@
+extern int16_t enc;
+extern int16_t encPrev;
+
+void housekeeping_task_user(void) {
```
wait, so you have the encoder wired up to a single pin, and then some manual behavior for the encoder push? 

If so, that's .... gross.  

While not necessarily better, having the encoder behavior in the normal matrix/layout would be better, IMO.  The work louder micro has a custom matrix with that sort of behavior, for reference.

RustedAperture replied:
@drashna I will have to investigate once I get the chance. This was an implementation that I grabbed from [lynepad](https://github.com/qmk/qmk_firmware/tree/042e42d1b7de603bc0a164b7d3b2ec761c6e140f/keyboards/qvex/lynepad). I'm working with my limited knowledge and was grabbing from other keyboards as an example. 

As much as I would like to get it right from the start, I prefer to work by examples since I'm unfamiliar with the code. 

I do agree that it was a dirty fix though and I will attempt to get it working.

drashna replied:
for reference, the work louder micro has 2 touch sensors that pull a digital pin high or low, depending on the status.  If the board was set up as "direct pin" these could have been used directly.  But since they weren't... a custom matrix was needed. Mostly, I took the normal matrix, removed everything that wasn't needed, and then added this block: 

https://github.com/qmk/qmk_firmware/blob/8bb7d24fba6c70dd5865b4d92421fcb9248b7f0b/keyboards/work_louder/micro/matrix.c#L99-L106

So on row 3, at col 0 and 3, it will read the pin for the touch sensors rather than the normal matrix read attempt.  Basically, that's what you want.  And then the normal keymap/layout can be used and you don't need a hard coded behavior.

RustedAperture replied:
@drashna I'm trying to get this to work for ROW2COL. I'm struggling to figure out where to put the check for row and col.

```
__attribute__((weak)) void matrix_read_rows_on_col(matrix_row_t current_matrix[], uint8_t current_col, matrix_row_t row_shifter) {    // Start with a clear matrix row
        bool key_pressed = false;

    // Select col
    if (!select_col(current_col)) { // select col
        return;                     // skip NO_PIN col
    }
    matrix_output_select_delay();

    // For each row...
    for (uint8_t row_index = 0; row_index < MATRIX_ROWS; row_index++) {
        // Check row pin state
        if (readMatrixPin(row_pins[row_index]) == 0) {
            // Pin LO, set col bit
            current_matrix[row_index] |= row_shifter;
            key_pressed = true;
        } else {
            // Pin HI, clear col bit
            current_matrix[row_index] &= ~row_shifter;
        }
    }

    // Unselect col
    unselect_col(current_col);
    matrix_output_unselect_delay(current_col, key_pressed); // wait for all Row signals to go HIGH
}
```

RustedAperture replied:
I've set the `matrix_init_custom` to be the following 
```
__attribute__((weak)) void matrix_init_custom(void) {
    unselect_cols();
    for (uint8_t x = 0; x < MATRIX_ROWS; x++) {
        if (row_pins[x] != NO_PIN) {
            setPinInputHigh_atomic(row_pins[x]);
        }
    }
    setPinInputHigh_atomic(B8);
}
```

RustedAperture replied:
Here is my `matrix.c` in its entirety 
```
// Copyright 2022 Cameron Varley (@RustedAperture)
// SPDX-License-Identifier: GPL-2.0-or-later

/*
 * scan matrix
 */
#include <stdint.h>
#include <stdbool.h>
#if defined(__AVR__)
#include <avr/io.h>
#endif
#include "wait.h"
#include "print.h"
#include "debug.h"
#include "util.h"
#include "matrix.h"
#include "debounce.h"
#include QMK_KEYBOARD_H

/* matrix state(1:on, 0:off) */
extern matrix_row_t matrix[MATRIX_ROWS];     // debounced values
extern matrix_row_t raw_matrix[MATRIX_ROWS]; // raw values

static const pin_t row_pins[MATRIX_ROWS] = MATRIX_ROW_PINS;
static const pin_t col_pins[MATRIX_COLS] = MATRIX_COL_PINS;
#define MATRIX_ROW_SHIFTER ((matrix_row_t)1)

static inline void setPinOutput_writeLow(pin_t pin) {
    ATOMIC_BLOCK_FORCEON {
        setPinOutput(pin);
        writePinLow(pin);
    }
}

static inline void setPinOutput_writeHigh(pin_t pin) {
    ATOMIC_BLOCK_FORCEON {
        setPinOutput(pin);
        writePinHigh(pin);
    }
}

static inline void setPinInputHigh_atomic(pin_t pin) {
    ATOMIC_BLOCK_FORCEON {
        setPinInputHigh(pin);
    }
}

static inline uint8_t readMatrixPin(pin_t pin) {
    if (pin != NO_PIN) {
        return readPin(pin);
    } else {
        return 1;
    }
}

static bool select_col(uint8_t col) {
    pin_t pin = col_pins[col];
    if (pin != NO_PIN) {
        setPinOutput_writeLow(pin);
        return true;
    }
    return false;
}

static void unselect_col(uint8_t col) {
    pin_t pin = col_pins[col];
    if (pin != NO_PIN) {
#            ifdef MATRIX_UNSELECT_DRIVE_HIGH
        setPinOutput_writeHigh(pin);
#            else
        setPinInputHigh_atomic(pin);
#            endif
    }
}

static void unselect_cols(void) {
    for (uint8_t x = 0; x < MATRIX_COLS; x++) {
        unselect_col(x);
    }
}


__attribute__((weak)) void matrix_init_custom(void) {
    unselect_cols();
    for (uint8_t x = 0; x < MATRIX_ROWS; x++) {
        if (row_pins[x] != NO_PIN) {
            setPinInputHigh_atomic(row_pins[x]);
        }
    }
    setPinInputHigh_atomic(B8);
}

__attribute__((weak)) void matrix_read_rows_on_col(matrix_row_t current_matrix[], uint8_t current_col, matrix_row_t row_shifter) {    // Start with a clear matrix row
        bool key_pressed = false;

    // Select col
    if (!select_col(current_col)) { // select col
        return;                     // skip NO_PIN col
    }
    matrix_output_select_delay();

    // For each row...
    for (uint8_t row_index = 0; row_index < MATRIX_ROWS; row_index++) {
        // Check row pin state
        if (readMatrixPin(row_pins[row_index]) == 0) {
            // Pin LO, set col bit
            current_matrix[row_index] |= row_shifter;
            key_pressed = true;
        } else {
            // Pin HI, clear col bit
            current_matrix[row_index] &= ~row_shifter;
        }
    }

    // Unselect col
    unselect_col(current_col);
    matrix_output_unselect_delay(current_col, key_pressed); // wait for all Row signals to go HIGH
}

bool matrix_scan_custom(matrix_row_t current_matrix[]) {
    static matrix_row_t temp_matrix[MATRIX_ROWS] = {0};

    matrix_row_t row_shifter = MATRIX_ROW_SHIFTER;
    for (uint8_t current_col = 0; current_col < MATRIX_COLS; current_col++) {
        matrix_read_rows_on_col(temp_matrix, current_col, row_shifter);
    }

    bool changed = memcmp(current_matrix, temp_matrix, sizeof(temp_matrix)) != 0;
    if (changed) {
        memcpy(current_matrix, temp_matrix, sizeof(temp_matrix));
    }
    return changed;
}
```

RustedAperture replied:
I want to map it to `{"label": "CALC", "matrix": [2, 3], "x": 3, "y": 2},`

RustedAperture replied:
I've added the matrix.c file in my most recent commit

RustedAperture replied:
@drashna I got it working now, please ignore above.

RustedAperture commented on keyboards/gmmk/numpad/matrix.c:
```diff
@@ -0,0 +1,133 @@
+    matrix_output_select_delay();
+
+    // For each row...
+    for (uint8_t row_index = 0; row_index < MATRIX_ROWS; row_index++) {
+        // Check row pin state
+        if (readMatrixPin(row_pins[row_index]) == 0) {
+            // Pin LO, set col bit
+            current_matrix[row_index] |= row_shifter;
+            key_pressed = true;
+        } else {
+            // Pin HI, clear col bit
+            current_matrix[row_index] &= ~row_shifter;
+        }
+    }
```
@drashna I'm not 100% sure how to add the readpin for the encoder push button on B8 and map it to 2,3 on the matrix here.

RustedAperture replied:
```suggestion
    for (uint8_t row_index = 0; row_index < MATRIX_ROWS; row_index++) {
        // Check row pin state
        if (current_col == 1 && row_index == 0 && readMatrixPin(B8) == 0) {
            current_matrix[row_index] |= row_shifter;
            key_pressed = !readMatrixPin(B8);
        } else if (readMatrixPin(row_pins[row_index]) == 0) {
            // Pin LO, set col bit
            current_matrix[row_index] |= row_shifter;
            key_pressed = true;
        } else {
            // Pin HI, clear col bit
            current_matrix[row_index] &= ~row_shifter;
        }
    }
```

RustedAperture replied:
But now the button is being spammed, not sure how to fix

drashna commented on keyboards/gmmk/numpad/matrix.c:
```diff
@@ -0,0 +1,136 @@
+ */
+#include <stdint.h>
+#include <stdbool.h>
+#if defined(__AVR__)
+#include <avr/io.h>
+#endif
```
```suggestion
```

drashna commented on keyboards/gmmk/numpad/matrix.c:
```diff
@@ -0,0 +1,136 @@
+#include "util.h"
+#include "matrix.h"
+#include "debounce.h"
+#include QMK_KEYBOARD_H
```
```suggestion
#include "quantum.h"
```
